src/common/summarize-profile.js

```diff
@@ -154,40 +174,59 @@ function sampleCategorizer(thread: Thread): SampleCategorizer {
     }
 
     const funcIndex = thread.frameTable.func[frameIndex];
-    const name = thread.stringTable._array[thread.funcTable.name[funcIndex]];
-    const category = categorizeFuncName(name);
-    if (category !== false && category !== 'wait') {
-      return category;
-    }
-
-    const prefixCategory = categorizeSampleStack(thread.stackTable.prefix[stackIndex]);
-    if (category === 'wait') {
-      if (prefixCategory === null || prefixCategory === 'uncategorized') {
-        return 'wait';
-      }
-      if (prefixCategory.endsWith('.wait') || prefixCategory === 'wait') {
-        return prefixCategory;
-      }
-      return prefixCategory + '.wait';
-    }
-
-    return prefixCategory;
+    return categorizeFuncName(funcIndex, thread.funcTable.name, thread.stringTable._array);
   }
 
-  const stackCategoryCache: Map<IndexIntoStackTable, (string|null)> = new Map();
-
   function categorizeSampleStack(stackIndex: (IndexIntoStackTable|null)): (string|null) {
     if (stackIndex === null) {
       return null;
     }
-    let category = stackCategoryCache.get(stackIndex);
-    if (category !== undefined) {
-      return category;
+
+    let category = 'uncategorized';
+    let lastWaitIndex;
+    let firstCategorizedIndex;
```

Comment:
The rest of this function has me totally confused. Maybe merging categorizeStack into this function would help. Its behavior is inconsistent in the cached vs non-cached cases: If cached, the result can be informed by the whole stack; if uncached, it only looks at the leaf function. I think this would confuse me less if categorizeStack were not a separate function.